### PR TITLE
Use ceph daemon-base as rook base

### LINKED ISSUE #1404: Consider using images from ceph-container project
With #1362 the Rook project has switched to using ceph Debian packages instead of building it from source. The goal is more alignment between Rook and upstream Ceph.

The image hierarchy is as follows:

```
Base Image (Ubuntu minified)
   -> Rook image (ceph installed from deb packages + rook binaries)
```

The `amd64` image ends up being ~180MB compressed with this approach.

The ceph-container project has also been working on size reductions in https://github.com/ceph/ceph-container/pull/877, and includes some of the same tricks to minify the base OS as Rook. The ceph-container project has the following hierarchy:

```
Base Ubuntu
   -> Ceph image (cleanup of OS + ceph installed from deb packages + etcd + confd etc.)
````

We should consider bringing both image together to something like:

```
Base Image (Ubuntu minified)
     Ceph Image (minimal ceph binaries installed from deb packages)
            Rook (rook binaries)
            or
            Ceph-Container (etcd + confd + scripts)
```

However for rook scenarios we would like to include other backends in our single image, and docker does not support multiple base images. If we were to add Gluster to the same Rook image, I think we would still do something like:

```
Base Image (Ubuntu minified)
     Rook Image (ceph and gluster installed from deb packages + rook binaries)
```

Alternatively Rook could use multiple images for each backend.

### LINKED ISSUE #1600: Consider Switching to  a less CVE havy baseimage
Hi, 

rook uses ubuntu/debian based docker images , if you take a look at the cve report from clair (cve sacnner from coreos)

there are overe 100 known unfixed CVEs

[analysis-registry.xxx.net-rook-v0.7.1_thirdparty.0.html.txt](https://github.com/rook/rook/files/1842104/analysis-registry.xxx.net-rook-v0.7.1_thirdparty.0.html.txt)

the most CVes are unfixed in Debian/Ubuntu

If you take a look at centos e.g there are 0 unfixed CVEs in the centos:7 image

Switching to centos:7 would be a big security improvment

### PR DESCRIPTION
**Pull request by BlaineEXE**
**DONE:** Wait to merge until after #1678 

Rook benefits from the testing the ceph-container project does, and the
scope of the rook project no longer has to maintain builds of base
images containing ceph packages.

The only arm builds of ceph-container are centos, and as switching to
centos was already a thought, that is done at the same time as this
change.

Resolves #1600 
Resolves #1404 

Checklist:
- [x] Documentation has been updated, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.

### CODE REVIEW

**Comment by BlaineEXE** [comment]
I have tested on my workstation that the crossbuild successfully builds amd64 (x86_64) and arm64 (aarch64) images. I updated the documentation to the best of my knowledge, but there is much that I may have missed. I have also updated the pending release notes, but let me know if the notes should be more appropriately moified.

**Comment by BlaineEXE** [comment]
@bassam @travisn :  The Jenkins build seems to be failing at a step "Populating the apt-get cache." I can't find a reference to such in the Rook source. I think I need some help figuring out what else may need to be changed here, if anything.

**Comment by travisn** [comment]
@BlaineEXE it looks like you're still in the middle of this big command: https://github.com/rook/rook/pull/1714/files#diff-64935596acae93da52032bce6ba2b928R18, but i'm not sure what the issue is

**Comment by BlaineEXE** [comment]
@bassam 
> another option is to leave the cross image on ubuntu.

I hadn't considered that. It seems to me that this might be the best option in an if-it-ain't-broke-don't-fix-it kind of way, but I could perhaps see an argument for using the same BaseOS as the rook images. Does anyone have any reasons that Ubuntu shouldn't be used?

**Comment by bassam** [comment]
yeah, less churn is better.

**Comment by BlaineEXE** [comment]
It turns out I wasn't building the cross container after all. I incorrectly assumed that because `make build.all` built both amd64 and arm64 images, the cross container was also built.

**Comment by BlaineEXE** [comment]
Thanks, @travisn for helping me to figure out that I had mistakenly based the toolbox just on centos:7 as a base rather than the rook image. All builds are now passing. 

As I mentioned above, I did my best to make the right changes to the documentation, but there is much of it. Let me know if I missed anything.

File: images/Makefile [hunk 1]
```diff
@@ -16,21 +16,19 @@ include image.mk
...
 
-# ====================================================================================
-# Image Targets
+export CEPH_BASE_IMAGE ?= ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7-$(PLATFORM_ARCH)
 
-apt-cacher: apt-cacher.linux_amd64
-apt-cacher.%: base.%
-	@$(MAKE) -C apt-cacher PLATFORM=$*
+# tini's version
+export TINI_VERSION = v0.16.1
```
Comment (bassam) [inline images/Makefile]: is this still needed?

File: images/Makefile [hunk 2]
```diff
@@ -16,21 +16,19 @@ include image.mk
 
 PLATFORMS ?= linux_amd64 linux_arm64
 
-# ====================================================================================
-# Image Targets
+export CEPH_BASE_IMAGE ?= ceph/daemon-base:v3.0.5-stable-3.0-luminous-centos-7-$(PLATFORM_ARCH)
```
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply images/Makefile]: Short of running the container and inspecting it, no.

I have opened up this issue against ceph-container that is related and to which I'd welcome your input: https://github.com/ceph/ceph-container/issues/1070

If the Rook team wants builds that have specific ceph versions, we (+ @leseb) can discuss building these and uploading them to the DockerHub in the ceph-container scripts. I think we might want to have a fairly detailed discussion about how often to update Ceph versions and if/how often to update the ceph-container code with that.

The above may be confusing ... I guess I mean that both ceph and ceph-container have versions that might want to be considered separately. Rook master I might suggest should have the latest ceph version (limited to major release like luminous) and latest ceph-container. Rook's versioned releases may want to have a fixed ceph version and fixed version of ceph-container, but those fixed versions should be updated at some interval to get security/bug fixes. Another approach could be to keep a fixed version of ceph-container but allow the ceph packages within to be updated to latest to get security/bug fixes more automatically. I see no real need to keep a fixed ceph version and allow ceph-container to always be updated to "latest," but this also is an option.

[the remark this answers is not in the document]
Comment (leseb) [reply images/Makefile]: IIRC ceph-container supports building a specific version of Ceph, we just don't do it at the moment. It's just for convenience if you really need it.

**Comment by BlaineEXE** [comment]
Something @travisn has pointed out is that "the uncompressed image size jumped from 434MB to 714MB."

I can theorize a couple possible reasons for this:
 - CentOS is the only release currently that has iSCSI packages, I believe. I'm not sure what the size of these packages are. (I would think that Rook eventually will want to configure iSCSI for Ceph, so this still likely belongs in the base image.)
 - yum packages don't have "suggests" or "recommends" package dependencies, only "yes/no" deps. It could be that unnecessary dependencies are pulled in by some recommended deps being given as "yes" deps in yum.

We do see that ceph-container's daemon-base image size is about 190MB larger for centos than for ubuntu/opensuse. 
```
ceph/daemon-base  suse-luminous-centos-7-x86_64       1f42d645fbc2  652MB
ceph/daemon-base  suse-luminous-opensuse-42.3-x86_64  48c744a489c2  464MB
ceph/daemon-base  suse-luminous-ubuntu-16.04-x86_64   9ceb3e14ee66  460MB
```

For Jewel and Kraken, however, the CentOS image is notably smaller than both. This may suggest that the iSCSI packages are a large part of this.
```
ceph/daemon-base  suse-jewel-centos-7-x86_64       37d25fdd81b9  579MB
ceph/daemon-base  suse-jewel-ubuntu-14.04-x86_64   6db29cd9626c  817MB
ceph/daemon-base  suse-jewel-ubuntu-16.04-x86_64   1affaa1ab026  748MB
ceph/daemon-base  suse-kraken-centos-7-x86_64      ce37cafe3565  600MB
ceph/daemon-base  suse-kraken-ubuntu-16.04-x86_64  39ef434a5fde  686MB
```

File: INSTALL.md [hunk 3]
```diff
@@ -149,7 +130,7 @@ Every PR and every merge to master triggers the CI process in [Jenkins](http://j
 The Jenkins CI will build, run unit tests, run integration tests and Publish artifacts- On every commit to PR and master.
 If any of the CI stages fail, then the process is aborted and no artifacts are published.
 On every successful build Artifacts are pushed to a [s3 bucket](https://release.rook.io/). On every successful master build,
-images are uploaded to quay and docker hub in addition.  
+images are uploaded to quay and docker hub in addition.
```
Comment (leseb) [inline INSTALL.md]: There is nothing on `quay` at the moment, we have the org, I'm just not sure if it's useful, @bassam, what do you think?

Comment (bassam) [reply INSTALL.md]: you don't need to push to quay. However, rook is currently pushing to quay.

**Comment by bassam** [comment]
Did you figure out what was causing the image bloat? 190MB seems like a lot.

**Comment by BlaineEXE** [comment]
Total mage size difference: 652MB - 460MB = 192MB

Base image size diff: 199MB - 113M = 86MB
Installed size of iSCSI packages as reported by yum: 46MB

There are still 60MB unaccounted for.
Ideas, @leseb or @dmick?

**Comment by BlaineEXE** [comment]
Also, it seems like other PRs may be seeing the same CI failure I am seeing here. Could this be a CI system issue, or am I missing something?

**Comment by dmick** [comment]
@BlaineEXE what two images are you comparing exactly?

**Comment by travisn** [comment]
@BlaineEXE yes we are having build issues lately, but you have a green build now.

**Comment by BlaineEXE** [comment]
> what two images are you comparing exactly?
@dmick , I am comparing the daemon-base images of ubuntu and centos (luminous).

**Comment by leseb** [comment]
@BlaineEXE I have no idea at the moment, I'm traveling this week so I don't think I'll have time to look into that sorry.

**Comment by BlaineEXE** [comment]
### Some size analysis:

After some input from @dmick, I've started looking at results from `du -h` of directories between ceph with centos base, ceph with ubuntu base, the centos base, and the ubuntu base images in a spreadsheet. I made the sheet editable to anyone with the link, so you can feel free to add to it if you like. 

https://docs.google.com/spreadsheets/d/1llRZjvX0H55lV-UN9Txs3wTmAw0jChTEMYd9EHgM5vA/edit?usp=sharing

What I'm seeing from my analysis of `/usr` so far is that the Ubuntu image grows slightly more than the CentOS image, but the CentOS base image is larger to start.

**Comment by f0** [comment]
Regarding the Image Size, see this Discusion[1] , and tis script[2] for buildinga ssmall (~60MB) base image with Buildah
[1]https://github.com/projectatomic/buildah/issues/532
[2]https://gitlab.com/pixdrift/buildah-scripts/blob/master/el7.minimal.sh

**Comment by jbw976** [comment]
I've opened issue #1742 to track reducing the image size

**Pull request merged**